while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+
+<br />
+
+Be creative!
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+### A Little Helper for the Next Question
+
+The following coding question uses a helper function called `input_int`. It works just like the [`input`](https://docs.python.org/3/library/functions.html#input) function that we've seen before, except that instead of returning a string, we will assume that it returns an integer.
+
+In the question block below, we don't need to define `input_int`. We can assume that it is available, the same way `input` is available.
+
+To experiment with the coding question in VS Code or repl.it, we will need an implementation for `input_int`. In Python, safely converting from a string to an integer requires handling exceptions. We haven't discussed that material yet, so here is a working implementation.
+
+<details style="max-width: 700px; margin: auto;">
+    <summary>A working implementation of `input_int`</summary>
+```python
+# accept a variable number of positional parameters
+def input_int(*args):
+    # loop forever (until we get valid input)
+    while True:
+        # get string input
+        user_input = input(*args)
+        try:
+            # try converting to an int (will return if successful)
+            return int(user_input)
+
+            # if the conversion succeeded, we will have returned already
+        except ValueError:
+            # could not convert the input, so show an error message
+            print('Try again.')
+```
+</details>
+
+
+<!-- prettier-ignore-start -->
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 0fdaf2f7-0573-4e0a-8fd2-46f15076de24
+* title: While Loops
+
+##### !question
+
+Let's sum some numbers! We will write a function called `silly_sum` that reads a series of numbers from user input and returns the sum. To keep things interesting, let's also decide that entering `0` means the user is done entering input, and that we'll also stop asking for further input if the sum reaches or exceeds `1000`.
+
+Provide an implementation for `silly_sum` so that it
+- reads numbers from the user (use `input_int`) summing as we go until either
+    - the user enters `0`, or
+    - the sum reaches or exceeds `1000`
+
+##### !end-question
+
+##### !placeholder
+
+```python
+def silly_sum():
+    # Be sure to call input_int when you need an integer from the user.
+    # A default implementation is provided above for experimenting in VS Code or repl.it.
+    # You should not paste that implementation here. It will be available automatically.
+    pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```python
+import unittest
+import main
+from main import silly_sum
+
+num_src = None
+
+def input_ints(nums):
+    for num in nums:
+        yield num
+
+def input_int(*args):
+    return next(num_src)
```

Comment:
Idk if there's something actionable to do here, but I'd love to call out that while I was debugging my solution to this, my stacktrace looked like this, which was specifically _not_ helpful to me lol.

```
.EE
======================================================================
ERROR: test_three (test_methods.TestPython1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/app/test_methods.py", line 37, in test_three
    self.assertEqual(1000, silly_sum(), 'the provided input was 999, 1, 1')
  File "/usr/src/app/main.py", line 9, in silly_sum
    user_input = input_int()
  File "/usr/src/app/test_methods.py", line 17, in input_int
    return next(num_src)
StopIteration

======================================================================
ERROR: test_two (test_methods.TestPython1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/app/test_methods.py", line 32, in test_two
    self.assertEqual(1200, silly_sum(), 'the provided input was 400, 400, 400, 400')
  File "/usr/src/app/main.py", line 9, in silly_sum
    user_input = input_int()
  File "/usr/src/app/test_methods.py", line 17, in input_int
    return next(num_src)
StopIteration

----------------------------------------------------------------------
Ran 3 tests in 0.001s

FAILED (errors=2)
```

Reply:
The test is simulating user input, and the code didn't stop reading it before reaching the end, which results in an uncaught StopIteration error.  The snowman project ran into this as well...

I _could_ add an except and translate it into a custom error that maybe reads more intuitively as a YouReadTooManyThingsTryAddingTerminateConditionsError (not this seriously, but something more descriptive than the built int StopIteration).

Thoughts?

Reply:
I personally feel like I don't have enough data to think of the "perfect" solution...

As a "sick and twisted" educator, instead of fool-proofing this, I think I propose:

- Doing a watered-down "catch all" hint: Make a hint that says "If you see a `StopIteration` error, it is because ... . This is likely happened because the function's logic did not stop before reaching the end of the list. To resolve this error, look for logical errors again. Try pairing and rubber duck debugging if you need to!"
- Possibly adding the smallest `.instructor.md` resource here and make a 2 second explanation about this (aka acknowledge that `StopIteration` error could exist)

Hopefully by C16, we'll have collective wisdom for how to _actually_ deal with this? Maybe?


Happy for other suggestions, too :D The suggestion I just gave is essentially "I don't know how to suggest to rewrite the question" looool


PS: For more data, I got this error because I checked for the number to be greater than `1000`, instead of the sum... purely a logical error, not syntax one!

Reply:
The hint tag is currently used to show the possible inputs that the tests are looking for, so I think I'd prefer to try raising a more descriptive error first.  If it doesn't work well, I'll fall back on sticking more info in the hint.

An instructor file just in case sounds good too. That'll be my first one!

Reply:
Now if you go past the end of input, you will get
`test_methods.ReadTooManyInputsError: Your loop has read beyond the end of the user input!`

While I do like the thought of having students dig into investigating the StopIteration, in all fairness, libraries _do_ try to report _useful_ exceptions rather than making us dig into their implementations.  If we were covering generators, then StopIteration would be great, but since it's really an implementation detail here, and students will have already slogged through a pretty long set of questions, I feel like taking it down a notch and just saying what really went wrong is reasonable.